sound/soc/sof/topology.c

```diff
@@ -1429,7 +1442,8 @@ static int sof_connect_dai_widget(struct snd_soc_component *scomp,
 }
 
 static int sof_widget_load_dai(struct snd_soc_component *scomp, int index,
-			       struct snd_sof_widget *swidget, int core,
+			       struct snd_sof_widget *swidget,
+			       struct sof_ipc_comp *comp,
```

Comment:
Maybe you could use the opportunity to add a bit more safety:
```C
const struct sof_ipc_comp * const comp;
```

Reply:
> Maybe you could use the opportunity to add a bit more safety:
> 
> ```c
> const struct sof_ipc_comp * const comp;
> ```

So do you suggest to do similar changes to all other arguments?

Reply:
Maybe, but not in this series/PR.

Reply:
for better or worse we don't often declare variables, including parameters, as `const`. In principle I certainly see value in that. If you just want a symbolic name for a constant but it's only used once in a function, so, it doesn't deserve a `#define`, you can do
```
void f(struct mydev *d)
{
	const u32 interrupt_enable = BIT(2);
	writel(interrupt_enable, d->config_reg);
}
```
and we do use this sometimes. But for function parameters it's way too often (almost always) the case, that the caller supplies a pointer to some context, which the function has no interest to change. Almost always such parameters are read-only. But we never use them, nor does the kernel in general seem to use them. So the second `const` doesn't really make much sense there. E.g. even without looking at the function I'm pretty sure other its pointer parameters are `const` too. One could even argue that function parameters could be `const` by default and the compiler could warn if they were written to.

Reply:
> for better or worse we don't often declare variables, including parameters, as `const`. In principle I certainly see value in that. If you just want a symbolic name for a constant but it's only used once in a function, so, it doesn't deserve a `#define`, you can do
> 
> ```
> void f(struct mydev *d)
> {
> 	const u32 interrupt_enable = BIT(2);
> 	writel(interrupt_enable, d->config_reg);
> }
> ```
> 
> and we do use this sometimes. But for function parameters it's way too often (almost always) the case, that the caller supplies a pointer to some context, which the function has no interest to change. Almost always such parameters are read-only. But we never use them, nor does the kernel in general seem to use them. So the second `const` doesn't really make much sense there. E.g. even without looking at the function I'm pretty sure other its pointer parameters are `const` too. One could even argue that function parameters could be `const` by default and the compiler could warn if they were written to.

Makes sense to me, thanks for the enlightening summary.

Reply:
> So the second const doesn't really make much sense there.

The second const is only to catch unintended mistakes inside the function. As we know kernel developers never make any stupid mistake, that's why it's not needed :-)

The first const is part of the API "contract", catches mistakes much more subtle and harder to debug and it is used A LOT in the kernel, just `git grep const`. If correct then the first const should not be omitted.

> One could even argue that function parameters could be const by default and the compiler could warn if they were written to.

Yes, that's what some newer and safer languages do: immutable by default.

Reply:
@marc-hb I didn't propose to omit the first `const` - I fully agree, that if it is indeed correct there, it should be used, it's often used in the kernel by others and by myself too. That's indeed a "good thing to have" (TM). It's the second `const` that I'm objecting to.